python/feathub/dsl/ast_type_derive_utils.py

```diff
@@ -0,0 +1,224 @@
+#  Copyright 2022 The Feathub Authors
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      https://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+from typing import Dict, List, Callable, Optional
+
+from feathub.common.exceptions import FeathubExpressionException
+from feathub.common.types import (
+    DType,
+    Bool,
+    Int32,
+    Float64,
+    String,
+    Bytes,
+    Int64,
+    Float32,
+    Timestamp,
+)
+from feathub.dsl.ast import (
+    ExprAST,
+    BinaryOp,
+    UminusOp,
+    CompareOp,
+    ValueNode,
+    FuncCallOp,
+    VariableNode,
+    ArgListNode,
+    CastOp,
+    LogicalOp,
+    GroupNode,
+    IsOp,
+    CaseOp,
+    NullNode,
+)
+
+TYPE_PRECISION_RANK: List[DType] = [Float64, Float32, Int64, Int32]
+
+# A map from the Feathub built-in function name to a callable that is used to
+# evaluate the result data type of the function. The callable takes a list of
+# data types of arguments and returns the data type of the result.
+FUNCTION_RESULT_TYPE_EVALUATOR: Dict[str, Callable[[List[DType]], DType]] = {
```

Comment:
We probably need a dedicated class to define all FeatHub built-in functions similar to the existing `AggFunc` class. Then we can put the function signature definition there and apply this definition across all processors, and use dedicatedd IT tests for these functions similar to how we test FeatureView subclasses for all processors.

Maybe add it as TODO if it is not easy to do in this PR?

Reply:
I agree adding a class for all Feathub built-in functions is necessary. And I think it should account for both built-in functions and user-defined functions (#72, #83). I added a TODO and issue #86 to keep track of it.